src/ssl/bio.cc

```diff
@@ -300,15 +312,19 @@ Ssl::ServerBio::readAndGive(char *buf, const int size, BIO *table)
 int
 Ssl::ServerBio::readAndParse(char *buf, const int size, BIO *table)
 {
-    const int result = readAndBuffer(table);
+    const int result = readAndBuffer(table, size);
     if (result <= 0)
         return result;
 
     try {
         if (!parser_.parseHello(rbuf)) {
             // need more data to finish parsing
             BIO_set_retry_read(table);
+            #if USE_OPENSSL_KTLS
+            return giveBuffered(buf, size);
```

Comment:
The old/official code suggests that we do not want to give TLS libraries (whatever they are) the raw bytes until we parse them.  PR code does not seem to match that logic. Could you please explain, in a GitHub comment (for now), why we change the logic here if USE_OPENSSL_KTLS?

Reply:
Reverted: this change was necessary at the early stage of prototyping to successfully establish the connection, but it seems to be no longer required.

https://github.com/squid-cache/squid/pull/1629/commits/d279eaec8649a6d293e6ff5687ac560d86a360d9

Reply:
Just reverting this may not be a good idea.

When using KTLS, perhaps the upper layer OpenSSL function (SSL_*) does not want this Ssl::ServerBio::read function to read more data from the socket (file descriptor) than the OpenSSL function is expecting because at some specific point the OpenSSL function changes the socket option.

https://github.com/openssl/openssl/blob/5963aa8c196d7c5a940a979299a07418527932af/include/internal/ktls.h#L70C3-L70C3


I cannot reproduce the issue that I was encountering while prototyping though, depending on the timing of packet arrival, the current (old/official) code potentially read more than expected data from the socket, which has been OK for non-ktls session but not for ktls.

Perhaps we have to consider some alternative way.

Reply:
Just reverting this may not be a good idea.

When using KTLS, perhaps the upper layer OpenSSL function (SSL_*) does not want this Ssl::ServerBio::read function to read more data from the socket (file descriptor) than the OpenSSL function is expecting because at some specific point the OpenSSL function changes the socket option.

https://github.com/openssl/openssl/blob/5963aa8c196d7c5a940a979299a07418527932af/include/internal/ktls.h#L70C3-L70C3


I cannot reproduce the issue that I was encountering while prototyping though, depending on the timing of packet arrival, the current (old/official) code potentially read more than expected data from the socket, which has been OK for non-ktls session but not for ktls.

Perhaps we have to consider some alternative way.

Reply:
https://github.com/squid-cache/squid/pull/1629/commits/ba88d4a6af19628e74a966c25897cb52e71b4c68

This commit is obviously not for release, but (I guess) something to limit the read size like this would be needed to ensure ktls establishment if keeping the logic (we do not want to give TLS libraries the raw bytes until we parse them) is mandatory. (Is it for DoS attack prevention or something?)

Reply:
> Is it for DoS attack prevention or something?

Unfortunately, I do not know the answer. I can try to research it, but it will take time. Right now, I can speculate that the primary goal here is _not_ attack prevention but something like making sure we can still splice the TCP connection (i.e. preventing OpenSSL _sending_ anything on the socket) if necessary. It is also possible that we wanted to finish configuring OpenSSL session state (using parsed Hello information and squid.conf settings) as necessary.

Without knowing what this old code does (or was supposed to do), I cannot fully review your changes, including the sketch in ba88d4a6af19628e74a966c25897cb52e71b4c68.

Reply:
I've changed the code to PEEK the socket while getting Hello information.
https://github.com/squid-cache/squid/pull/1629/commits/d1c9d68d79f67154c1048c95b6f8f114b4b96f38

Hopefully this satisfies both of correct ktls establishment and "not-giving the raw bytes to the upper layer till Hello."

Actually it seems to be working on FreeBSD, however, I was not sure if it also works on every platform that squid supports so reintroduced the build switch macro in the later commits.